app/letter_attachment/rest.py

```diff
@@ -41,3 +43,21 @@ def create_letter_attachment():
     dao_update_template(template)
 
     return jsonify(letter_attachment.serialize()), 201
+
+
+@letter_attachment_blueprint.route("/letter-attachment/<uuid:letter_attachment_id>/archive", methods=["POST"])
+def archive_letter_attachment(letter_attachment_id):
+    data = request.get_json()
+
+    validate(data, post_archive_letter_attachment_schema)
+
+    letter_attachment = LetterAttachment.query.get_or_404(letter_attachment_id)
+
+    if (template := dao_get_template_by_id(data["template_id"])) is None:
+        raise InvalidRequest("template-not-found", 404)
+    template.letter_attachment = None
```

Comment:
I think we should rewrite this check so that we don't need to post the template ID across.

We can just use `letter_attachment.temple` to find the related template. If it exists then it's not been archived; if it doesn't then it's already been archived.

This protects us from another little potential edge case here - someone could send any old template_id across and we'll wipe the letter attachment from it. In this implementation we aren't enforcing that the letter_attachment being archived is attached to the template we're removing the attachment from.

Using the relationship takes care of this check for us.

I'd also be interested in your thoughts about 400 vs 404 here. I'm tempted to say that trying to archive an attachment that has already been archived should be a 400 rather than a 404, since we aren't _actually_ deleting it the resource does exist - but it's not valid to try to archive it again.

It depends how much we want to pretend that the underlying resource is being deleted rather than just hidden.

Reply:
I've updated to follow your suggestion. I seemed to have to look it up by the DAO, as It was causing this error.
`Can't record history for Template (something in your code has casued the database to flush the session early so there's nothing to copy into the history table)`
But it is using the id from the letter attachment relationship to do the lookup. I'll dig into the versioning thing a bit.

The 404 vs 400. I don't think we'll see many of these in practice. Unless someone is doing something hinkey to our API I'd expect it to only show up if someone was managing to press the button twice. I think 400 makes a little more sense, since you are making a bad request given the context. I've switched it over.